discovery/README.md

```diff
@@ -0,0 +1,123 @@
+lookups). No other labelnames should be exposed.
+
+It is very common for initial pull requests for new SDs to include hardcoded
+assumptions that make sense for the the author's setup. SD should be generic,
+any customisation should be handled via relabelling. There should be basically
+no business logic, filtering, or transformations of the data from the SD beyond
+that which is needed to fit it into the metadata data model. 
+
+Arrays (e.g. a list of tags) should be converted to a single label with the
+array values joined with a comma. Also prefix and suffix the value with a
+comma. So for example the array `[a, b, c]` would become `,a,b,c,`. As
+relabelling regexes are fully anchored, this makes it easier to write correct
+regexes against (`.*,a,.*` works no matter where `a` appears in the list). The
+canonical example of this is `__meta_consul_tags`.
+
+Maps, hashes and other forms of key/value pairs should be all prefixed and
+exposed as labels. For example for EC2 tags, there would be
+`__meta_ec2_tag_Description=mydescription` for the Description tag. Labelnames
+may only contain `[_a-zA-Z0-9]`, sanitize by replacing with underscores as needed.
+
+For targets with multiple potential ports, you can a) expose them as a list, b)
+if they're named expose them as a map or c) expose them each as their own
+target. Kuberenetes SD takes the target per port approach. a) and b) can be
+combined.
+
+For machine-like SDs (OpenStack, EC2, Kubernetes to some extent) there may
+be multiple network interfaces for a target. Thus far reporting the details
+of only the first network interface has sufficed.
+
+
+### Other implementation considerations
+
+SDs are intended to dump all possible targets. For example the optional use of
+EC2 service discovery would be to take the entire region's worth of EC2
+instances it provides and do everything needed in one `scrape_config`. For
+large deployments where you are only interested in a small proportion of the
+returned targets, this may cause performance issues. If this occurs it is
+acceptable to also offer filtering via whatever mechanisms the SD exposes. For
+EC2 that would be the `Filter` option on `DescribeInstances`. Keep in mind that
+this is a performance optimisation, it should be possible to do the same
+filtering using relabelling alone. As with SD generally, we do not invent new
+ways to filter targets (that is what relabelling is for), merely offer up
+whatever functionality the SD itself offers.
+
+It is a general rule with Prometheus that all configuration comes from the
+configuration file. While the libraries you use to talk to the SD may also
+offer other mechanisms for providing configuration/authentication under the
+covers (EC2's use of enviroment variables being a prime example), using your SD
+mechanism should not require this. Put another way, your SD implementation
+should not read enviroment variables or files to obtain configuration.
```

Comment:
I didn't want to get into too much depth here, I tried the capture what I think really matters to us (config is in the config file) while allowing for the various things the AWS library does internally and we seem nominally okay with.
The ARN thing is basically a form of sudo, so that's something different again.